autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py

```diff
@@ -371,7 +458,13 @@ def early_stop_handler(self, X: Dict[str, Any]) -> bool:
         if self.checkpoint_dir is None:
             self.checkpoint_dir = tempfile.mkdtemp(dir=X['backend'].temporary_directory)
 
-        epochs_since_best = self.run_summary.get_last_epoch() - self.run_summary.get_best_epoch()
+        if X['val_indices'] is None:
+            if X['X_test'] is not None:
+                epochs_since_best = self.run_summary.get_last_epoch() - self.run_summary.get_best_epoch('test_loss')
+            else:
+                epochs_since_best = self.run_summary.get_last_epoch() - self.run_summary.get_best_epoch('train_loss')
+        else:
+            epochs_since_best = self.run_summary.get_last_epoch() - self.run_summary.get_best_epoch()
```

Comment:
It is a bit shorter and clear.

```suggestion
        target_metrics = None  # If None, we will count from the epoch of the best validation
        if X['val_indices'] is None:
            target_metrics = 'test_loss' if X['X_test'] is not None else 'train_loss'

        epochs_since_best = self.run_summary.get_last_epoch() - self.run_summary.get_best_epoch(target_metrics)
```